### ISSUE TITLE
403 error on connection to Minio with AWS Signature V4

### ISSUE THREAD

**Issue by roby255** [caddyserver/caddy#3952]
Hi guys,

I run my CaddyServer v2.3 on Windows 10.
Caddy proxy the request for a Minio server running on port 9000.

I have some issues to connect to minio domain with AWS Signature Version 4, with the client QNAP hybrid backup
and IPERIUS BACKUP.
If I use the S3 Browser as client, it connect with Signature V4 without issues.

Before to install Caddy as proxy I had no issues connecting directly to the port with also let's encrypt SSL installed.

I tryed to connect with Signature V2 with no issue... the problem is only with Signature V4.
I checked my time synchronization on both client and server side and appear to be ok.

this is my Caddyfile configuration:
```
cloud1.robyho.it {
    reverse_proxy localhost:9000 {
        header_up X-Forwarded-Proto {scheme}
        header_up X-Forwarded-Host {host}
        header_up Host {host}
    }
	log {
		output file C:\SpettacoloUtility\Caddy\logs\cloud1.log {
			roll true				# Rotate logs, enabled by default
			roll_size_mb 5			# Set max size 5 MB
			roll_gzip true			# Whether to compress rolled files
			roll_local_time true	# Use localhost time
			roll_keep 2				# Keep at most 2 log files
			roll_keep_days 7		# Keep log files for 7 days 
		}
	}
}
```

this is error log on QNAP hybrid backup connection:
`{"level":"error","ts":1609621309.814594,"logger":"http.log.access.log1","msg":"handled request","request":{"remote_addr":"3.47.281.225:52090","proto":"HTTP/1.1","method":"GET","host":"example.it","uri":"/","headers":{"Authorization":["AWS4-HMAC-SHA256 Credential=some_id/20210102/Cloud1/s3/aws4_request,SignedHeaders=host;user-agent;x-amz-content-sha256;x-amz-date,Signature=8f2c649b94fe34fb9494ae78caf3ad09e6fca6fc654888ac87e23298414bd452"],"Accept-Encoding":["identity"],"User-Agent":["HybridBackup/14.0.1215 (QNAP; QTS 4.5.1)"],"X-Amz-Date":["20210102T210146Z"],"X-Amz-Content-Sha256":["e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"]},"tls":{"resumed":false,"version":772,"cipher_suite":4865,"proto":"","proto_mutual":true,"server_name":"example.it"}},"common_log":"3.47.281.225 - - [02/Jan/2021:22:01:49 +0100] \"GET / HTTP/1.1\" 403 357","duration":0.0010006,"size":357,"status":403,"resp_headers":{"Server":["Caddy","MinIO"],"Content-Type":["application/xml"],"X-Amz-Bucket-Region":["cloud1"],"X-Xss-Protection":["1; mode=block"],"X-Amz-Request-Id":["1656860D9A569538"],"Content-Security-Policy":["block-all-mixed-content"],"Vary":["Origin"],"Accept-Ranges":["bytes"],"Date":["Sat, 02 Jan 2021 21:01:49 GMT"],"Content-Length":["357"]}}
`

this is info log on S3 Browser connection:
`{"level":"info","ts":1609623722.3120239,"logger":"http.log.access.log1","msg":"handled request","request":{"remote_addr":"3.47.281.225:62727","proto":"HTTP/1.1","method":"GET","host":"example.it","uri":"/","headers":{"Connection":["Keep-Alive"],"User-Agent":["S3 Browser 9.2.1 https://s3browser.com"],"Authorization":["AWS4-HMAC-SHA256 Credential=some_id/20210102/cloud1/s3/aws4_request,SignedHeaders=host;user-agent;x-amz-content-sha256;x-amz-date, Signature=ce453bdd279263ac55a9faf95fde8a45a2cd250ead24a9ad2b3ffa73664461de"],"X-Amz-Content-Sha256":["e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"],"X-Amz-Date":["20210102T214202Z"]},"tls":{"resumed":false,"version":771,"cipher_suite":49196,"proto":"","proto_mutual":true,"server_name":"example.it"}},"common_log":"3.47.281.225 - - [02/Jan/2021:22:42:02 +0100] \"GET / HTTP/1.1\" 200 581","duration":0.0009986,"size":581,"status":200,"resp_headers":{"Date":["Sat, 02 Jan 2021 21:42:02 GMT"],"X-Amz-Request-Id":["1656883F4E640294"],"X-Xss-Protection":["1; mode=block"],"Content-Length":["581"],"Content-Type":["application/xml"],"Server":["Caddy","MinIO"],"Vary":["Origin"],"X-Amz-Bucket-Region":["cloud1"],"Accept-Ranges":["bytes"],"Content-Security-Policy":["block-all-mixed-content"]}}`

It's possible that Caddy transfer the useragent header (o one of the other header) in a wrong way? Could by one of the special character in that header (ex. HybridBackup/14.0.1215 (QNAP; QTS 4.5.1)?
The only difference that I see is the UserAgent Header that is used in X-Amz-SignedHeaders to calculate X-Amz-Signature.
[Authenticating Requests: Using Query Parameters (AWS Signature Version 4)](https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html)

Thank you for your work.
And sorry for my bad english

Resource:
[Setup Caddy proxy with MinIO Server](https://docs.minio.io/docs/setup-caddy-proxy-with-minio.html)
[Authenticating Requests (AWS Signature Version 4)](https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-authenticating-requests.html)
[S3 browser](https://s3browser.com/)

**Comment by mholt** [comment]
Can you distill this down a little more? For example, what is the value of the User-Agent header that Caddy is sending to the backend, versus what you expect it to be?

**Comment by roby255** [comment]
I made a mistake.

@mholt  after your message I used Fiddler to monitor the http packet before and after Caddy.
Look always the same.

Then I tryed again to connect directly Minio escluding Caddy, in two way:
- Without SSL: it work like a charm with Signature V4
- With Let's encrypt SSL (directly in Minio): I have the same problem with Signature V4

So I checked again the QNAP Hybrid Backup Sync configuration, and I saw that the firt letter of "server address" was a capital letter...
This create the problem in QNAP Hybrid Backup Sync when SSL is enabled and with Signature V4. So I made a correction and now it work.

I don't understand why Iperius backup still doesn't work with Signature V4... there's no digit error there.
The problem was not due to Caddy I think.

Now I can say that Caddy work like a charm with Minio, also under Cloudflare Proxy (I didn't mention it because all the tests were done with cloudflare disabled)

Thank you for all your work and sorry for my wrong report!
Best regards.

Resource:
[fiddler (http packet tracer)](https://www.telerik.com/fiddler)
